src/OSPSuite.Core/Domain/Builder/SimulationBuilder.cs

```diff
@@ -85,15 +93,102 @@ public IReadOnlyList<string> AllPresentEndogenousStationaryMoleculeNames() =>
+         mergeInitialConditions();
          cacheEntities();
       }
 
+      private void mergeBuilders<T>(Func<Module, IBuildingBlock<T>> propAccess, CacheByName<T> cache, Action<T, BuilderSource<T>> mergeStrategyAction) where T : class, IBuilder, IEntity
+      {
+         var analyzedMerges = analyzeBuilderMerges(propAccess);
+
+         foreach (var mergeInfo in analyzedMerges)
+         {
+            var (finalBuilder, buildingBlock) = mergeInfo.BaseBuilder;
+
+            if (mergeInfo.RequiresClone)
+            {
+               finalBuilder = cloneBuilder(finalBuilder);
+               foreach (var (sourceBuilder, sourceBuildingBlock) in mergeInfo.BuildersToMerge)
+               {
+                  //Clone the source builder if needed to prevent cross-contamination during sequential merges
+                  //When merging multiple builders, earlier merged builders could be affected by later merges through shared references
+                  var sourceBuilderToMerge = mergeInfo.RequiresMergeClone ? cloneBuilder(sourceBuilder) : sourceBuilder;
+                  var mergedBuilderSource = new BuilderSource<T>(sourceBuilderToMerge, sourceBuildingBlock);
+                  tryMergeContainers(finalBuilder, mergedBuilderSource);
+                  mergeStrategyAction(finalBuilder, mergedBuilderSource);
+               }
+            }
+
+            cache.Add(finalBuilder);
+            AddToBuilderSource(finalBuilder, buildingBlock);
+         }
+      }
+
+      private T cloneBuilder<T>(T builder) where T : class, IBuilder
+      {
+         //we will need to make sure we keep the Id the same somehow to ensure proper tracking of entities. TBD
+         return _cloneManager.Clone(builder);
+      }
+
+      private void tryMergeContainers<T>(T target, BuilderSource<T> builderSource) where T : IEntity
+      {
+         var targetContainer = target as IContainer;
+         var sourceContainer = builderSource.Builder as IContainer;
+         if (targetContainer == null || sourceContainer == null)
+            return;
+
+         mergeContainers(targetContainer, new BuilderSource<IContainer>(sourceContainer, builderSource.BuildingBlock));
+      }
+
+      private void mergeContainers<T>(T target, BuilderSource<T> source) where T : IContainer
+      {
+         var (sourceBuilder, sourceBuildingBlock) = source;
+         //Marking all entities in the source as coming from this source
+         var allEntities = sourceBuilder.GetAllChildren<IEntity>();
+         allEntities.Each(entity => AddToBuilderSource(entity, sourceBuildingBlock));
+
+         //at this step, all entities in the target should already be marked as coming from their respective source
+         _containerMergeTask.MergeContainers(target, sourceBuilder);
+      }
+
+      private void mergeReactions(ReactionBuilder target, BuilderSource<ReactionBuilder> source)
+      {
+         //TODO add reaction merge behavior
```

Comment:
@benjaperez1983 This is where the logic in reaction merge from #2640 should be implemented. Most of the code in this PR does not apply but we should talk about what can be saved anyways.